File: src/runtime/agent-pool.ts
import type { AgentState, RunnableAgent } from "./lifecycle";

// #EXT-003-REQ-2 Start
// #EXT-003-REQ-4 Start
/**
 * A factory that produces a fresh runnable agent to run. The pool calls it when
 * capacity is available. The factory must return a freshly-spawned unit (state
 * `spawned`); the pool drives its `run()`/`teardown()` lifecycle.
 */
export interface AgentFactory {
  (): RunnableAgent;
}

/** Observability record for a single active agent. */
export interface AgentSnapshotEntry {
  readonly id: string;
  readonly state: AgentState;
}

/** Reports a contained agent failure to the harness/owner. */
export interface OnAgentFailed {
  (failure: { readonly id: string; readonly error: unknown }): void;
}

/** Construction options for an {@link AgentPool}. */
export interface AgentPoolOptions {
  /** Reports each contained agent failure; siblings keep running regardless. */
  readonly onAgentFailed?: OnAgentFailed;
}

/**
 * A bounded, observable pool of lightweight agent threads.
 *
 * The pool hosts up to `bound` concurrent agents. {@link submit} applies
 * backpressure: once `active().length === bound`, further spawns are QUEUED
 * rather than allowed to grow unbounded; queued factories start as running
 * agents complete and free a slot.
 *
 * Fault containment: a single agent's failure is contained by the
 * {@link RunnableAgent} and reported via `onAgentFailed`; sibling agents are
 * unaffected and continue running, and a freed slot still admits queued work.
 */
export class AgentPool {
  private readonly bound: number;
  private readonly onAgentFailed?: OnAgentFailed;
  private readonly activeThreads = new Map<string, RunnableAgent>();
  private readonly waiting: Array<{
    factory: AgentFactory;
    resolve: (t: RunnableAgent) => void;
    reject: (err: unknown) => void;
  }> = [];

  constructor(bound: number, options: AgentPoolOptions = {}) {
    if (!Number.isInteger(bound) || bound < 1) {
      throw new TypeError("AgentPool bound must be a positive integer.");
    }
    this.bound = bound;
    this.onAgentFailed = options.onAgentFailed;
  }

  /** The configured maximum number of concurrent agents. */
  get capacity(): number {
    return this.bound;
  }

  /** The currently running/active agent threads. */
  active(): readonly RunnableAgent[] {
    return [...this.activeThreads.values()];
  }

  /** Number of submissions queued behind the bound (backpressure depth). */
  get pending(): number {
    return this.waiting.length;
  }

  /**
   * Submit an agent for execution. If there is free capacity the agent is
   * spawned immediately; otherwise the factory is queued (backpressure) and
   * admitted when a slot frees. Resolves with the {@link RunnableAgent} once it has
   * been admitted and started running.
   */
  submit(factory: AgentFactory): Promise<RunnableAgent> {
    if (this.activeThreads.size < this.bound) {
      return Promise.resolve(this.start(factory));
    }
    return new Promise<RunnableAgent>((resolve, reject) => {
      this.waiting.push({ factory, resolve, reject });
    });
  }

  /** Enumerate active agents' ids and states for observability. */
  snapshot(): AgentSnapshotEntry[] {
    return this.active().map((t) => ({ id: t.id, state: t.state }));
  }

  /**
   * Wait until all active and queued agents have finished and been torn down.
   * Useful for deterministic shutdown in tests and callers.
   */
  async drain(): Promise<void> {
    while (this.activeThreads.size > 0 || this.waiting.length > 0) {
      const running = this.active().map((t) => t.teardown());
      await Promise.all(running);
      // teardown -> run() resolution removes from the map and admits queued work,
      // but loop again to catch newly-admitted agents.
      await Promise.resolve();
    }
  }

  /** Admit one factory: spawn the thread, register it, and drive its run. */
  private start(factory: AgentFactory): RunnableAgent {
    const thread = factory();
    if (this.activeThreads.has(thread.id)) {
      throw new Error(`Agent "${thread.id}" is already active in the pool.`);
    }
    this.activeThreads.set(thread.id, thread);

    // Drive the agent. run() never rejects (faults are contained inside the
    // thread), so a single failure cannot reject up into the pool or siblings.
    void thread.run().then(() => {
      this.activeThreads.delete(thread.id);
      if (thread.state === "failed") {
        try {
          this.onAgentFailed?.({ id: thread.id, error: thread.error });
        } catch {
          // The failure reporter must not destabilize the pool or siblings.
        }
      }
      this.admitNext();
    });

    return thread;
  }

  /** Admit the next queued factory, if any, now that a slot has freed. */
  private admitNext(): void {
    if (this.waiting.length === 0) {
      return;
    }
    if (this.activeThreads.size >= this.bound) {
      return;
    }
    const next = this.waiting.shift()!;
    try {
      const thread = this.start(next.factory);
      next.resolve(thread);
    } catch (err) {
      next.reject(err);
    }
  }
}
// #EXT-003-REQ-4 End
// #EXT-003-REQ-2 End

File: jaros/runtime/agent_pool.py
"""The bounded, observable concurrent agent pool (EXT-003 / REQ-2, REQ-4).

The pool hosts many agents at once but never more than a configurable ``bound``
running concurrently. When the bound is reached, further spawn requests are
*queued* (backpressure) rather than allowed to grow unbounded; queued work is
admitted as running slots free up.

The pool is observable: :meth:`active` enumerates running agents and
:meth:`snapshot` reports every agent's id + state. Faults are contained per
:class:`~jaros.runtime.agent_thread.AgentThread`: a failing agent is reported via
``on_agent_failed`` while its siblings keep running, and :meth:`drain` reaps
everything deterministically.
"""

from __future__ import annotations

import threading
from collections import deque
from dataclasses import dataclass
from typing import Callable

from jaros.runtime.agent_thread import AgentThread, FailureHook
from jaros.runtime.lifecycle import AgentState

# A factory produces a fresh agent on demand. The pool installs its own failure
# hook around the agent, so factories need not wire one up themselves.
AgentFactory = Callable[[], AgentThread]


@dataclass(frozen=True)
class AgentSnapshot:
    """An observable (id, state) pair for one agent."""

    id: str
    state: AgentState


# #EXT-003-REQ-2 Start
# #EXT-003-REQ-4 Start
class AgentPool:
    """A bounded pool of concurrently running agent threads.

    Args:
        bound: Maximum number of agents running concurrently (must be >= 1).
        on_agent_failed: Optional hook fired with ``(agent, exception)`` when a
            hosted agent's body raises. Failures are contained; siblings and the
            pool keep running.
    """

    def __init__(
        self,
        bound: int,
        on_agent_failed: FailureHook | None = None,
    ) -> None:
        if bound < 1:
            raise ValueError("bound must be >= 1")
        self._bound = bound
        self._on_agent_failed = on_agent_failed
        self._lock = threading.RLock()
        self._running: dict[str, AgentThread] = {}
        self._queue: deque[AgentFactory] = deque()
        # Every agent the pool has admitted, retained for observability/draining.
        self._known: list[AgentThread] = []

    @property
    def bound(self) -> int:
        """The configured maximum number of concurrently running agents."""
        return self._bound

    @property
    def pending(self) -> int:
        """Number of spawn requests queued behind the bound (backpressure)."""
        with self._lock:
            return len(self._queue)

    def submit(self, factory: AgentFactory) -> None:
        """Submit an agent factory for execution.

        If a running slot is free the agent is spawned immediately; otherwise the
        request is queued and admitted later as slots free (backpressure — the
        pool never exceeds ``bound`` concurrently running agents).
        """
        with self._lock:
            if len(self._running) < self._bound:
                self._spawn(factory)
            else:
                self._queue.append(factory)

    def active(self) -> list[AgentThread]:
        """Return the agents currently running (a snapshot list)."""
        with self._lock:
            return list(self._running.values())

    def snapshot(self) -> list[AgentSnapshot]:
        """Return an (id, state) snapshot of every agent the pool has admitted."""
        with self._lock:
            return [AgentSnapshot(id=a.id, state=a.state) for a in self._known]

    def drain(self) -> None:
        """Run all queued work to completion and tear down every agent.

        Repeatedly admits queued factories as slots free, joins each running
        agent, and tears it down deterministically. Returns once nothing is
        running and the backlog is empty.
        """
        while True:
            with self._lock:
                running = list(self._running.values())
                pending = len(self._queue)
            if not running and pending == 0:
                break
            for agent in running:
                agent.teardown()  # joins; _reap (via the hook below) frees the slot
        # Every admitted agent has finished; tear each down so its handle is
        # released deterministically and its terminal state is recorded (an
        # agent reaped between drain passes may still need its teardown call).
        with self._lock:
            known = list(self._known)
        for agent in known:
            agent.teardown()

    # --- internals ----------------------------------------------------------

    def _spawn(self, factory: AgentFactory) -> None:
        """Build an agent from ``factory``, wire reaping, and start it.

        Caller must hold ``self._lock``.
        """
        agent = factory()
        # Wrap any user failure hook so the pool both reports and reaps on fault.
        user_hook = agent.on_failed

        def _on_failed(a: AgentThread, exc: BaseException) -> None:
            if user_hook is not None:
                user_hook(a, exc)
            if self._on_agent_failed is not None:
                self._on_agent_failed(a, exc)

        agent.on_failed = _on_failed
        # Reap when the body finishes (success or contained failure) to free the
        # slot and admit queued work, without the caller having to join. The
        # thread targets the wrapper directly so reaping is guaranteed even if
        # the factory pre-created a thread bound to the bare ``run``.
        wrapped = self._wrap_run(agent)
        thread = threading.Thread(
            target=wrapped, name=f"agent-{agent.id}", daemon=True
        )
        object.__setattr__(agent, "_thread", thread)

        self._running[agent.id] = agent
        self._known.append(agent)
        thread.start()

    def _wrap_run(self, agent: AgentThread) -> Callable[[], None]:
        """Return a run wrapper that reaps the agent once its body returns."""
        original_run = AgentThread.run

        def _run() -> None:
            try:
                original_run(agent)
            finally:
                self._reap(agent.id)

        return _run

    def _reap(self, agent_id: str) -> None:
        """Free the slot held by ``agent_id`` and admit one queued factory."""
        with self._lock:
            self._running.pop(agent_id, None)
            if self._queue and len(self._running) < self._bound:
                self._spawn(self._queue.popleft())
# #EXT-003-REQ-4 End
# #EXT-003-REQ-2 End
